python/mxnet/gluon/model_zoo/text/base.py

```diff
@@ -74,22 +81,50 @@ def get_rnn_cell(mode, num_layers, num_hidden,
             rnn_cell.add(cell)
             if i != num_layers - 1 and dropout != 0:
                 rnn_cell.add(rnn.DropoutCell(dropout))
+    
+    if weight_dropout:
+        if training:
+            _apply_weight_drop_to_rnn_cell(rnn_cell, rate = weight_dropout, mode = 'training')
+        else:
+            _apply_weight_drop_to_rnn_cell(rnn_cell, rate = weight_dropout, mode = 'always')
+    
     return rnn_cell
 
-def get_rnn_layer(mode, num_layers, num_embed, num_hidden, dropout, weight_dropout):
-    # TODO weight drop
+
+def _apply_weight_drop_to_rnn_layer(block, rate, weight_dropout_mode = 'training'):
+    params = block.collect_params('.*_h2h_weight')
+
+    for key, value in params.items():
+        weight_dropped_params = WeightDropParameter(value, rate, weight_dropout_mode)
+        block.collect_params('.*_h2h_weight')._params[key] = weight_dropped_params
+        for child_block in block._children:
+            child_block.collect_params('.*_h2h_weight')._params[key] = weight_dropped_params
+            
+        block._unfused.params._params.clear()
+        for _unfused_child_block in block._unfused:
+            _unfused_child_block.collect_params('.*_h2h_weight')._params[key] = weight_dropped_params
+
+def get_rnn_layer(mode, num_layers, num_embed, num_hidden, dropout, weight_dropout, weight_dropout_mode = 'training'):
     if mode == 'rnn_relu':
-        return rnn.RNN(num_hidden, 'relu', num_layers, dropout=dropout,
-                       input_size=num_embed)
+        block = rnn.RNN(num_hidden, 'relu', num_layers, dropout=dropout,
+                       input_size=num_embed)      
     elif mode == 'rnn_tanh':
-        return rnn.RNN(num_hidden, num_layers, dropout=dropout,
+        block = rnn.RNN(num_hidden, num_layers, dropout=dropout,
                        input_size=num_embed)
     elif mode == 'lstm':
-        return rnn.LSTM(num_hidden, num_layers, dropout=dropout,
+        block = rnn.LSTM(num_hidden, num_layers, dropout=dropout,
                         input_size=num_embed)
     elif mode == 'gru':
-        return rnn.GRU(num_hidden, num_layers, dropout=dropout,
+        block = rnn.GRU(num_hidden, num_layers, dropout=dropout,
                        input_size=num_embed)
+    elif mode == 'awd-lstm':
+        block = rnn.LSTM(num_hidden, num_layers, dropout=dropout,
+                        input_size=num_embed)
+        if weight_dropout:
+            _apply_weight_drop_to_rnn_layer(block, weight_dropout, weight_dropout_mode = weight_dropout_mode)
```

Comment:
This function is a utility function for constructing RNN layers with a set of tricks. It doesn't need to know whether the block will be in awd or not. In that sense, you may move the weight_drop block outside of the last if condition and remove the awd-specific mode.

Reply:
The set of tricks is only for lstm, so I restrict the awd option only for lstm. We may want to hide the details to the model users, because we also do not the effectiveness of the model on other RNNs. It might cause unnecessary confusion for the users. How do you think?